### PR TITLE
Implement substrate api sidecar

### LINKED ISSUE #4974: Test substrate api sidecar HTTP RPC
**Is your feature request related to a problem? Please describe.**

REST API service intended to run next to Substrate, exposing a limited set of endpoints over HTTP with meaningful responses:

https://github.com/paritytech/substrate-api-sidecar

Doc:

https://paritytech.github.io/substrate-api-sidecar/dist

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by mariopino**
Closes #4974 

**Description:**

Implement substrate api sidecar replacing as much as possible polkadot-js api queries.

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by Bitcoinera** [comment]
Mario, if you merge with develop the commits from Fabo's PR will disappear, since this was already merged

File: api/lib/source/polkadotV0-source.js [hunk 1]
```diff
@@ -110,43 +114,75 @@ class polkadotAPI {
...
+    //   blockHash,
+    //   sessionIndex.toNumber(),
+    //   author,
+    //   transactions,
+    //   data
+    // )
+
+    let block
+    if (blockHeight) {
+      block = await this.get(`${this.baseURL}/block/${blockHeight}`)
```
Comment (faboweb) [inline api/lib/source/polkadotV0-source.js:162]: looks promising!

**Comment by mariopino** [comment]
We can use parameters like this:

```
/pallets/staking/storage/nominators?key1=Ckus4fsdJHxVg69FMDPVbqre3cLUc5R6fWxkMR1HuyQC1Dk
```

```
{
   "at":{
      "hash":"0x212a47ee43b076a76e281b401f1c8e316b82f56c3f319b1fcc0322ccc54d9af9",
      "height":"4162539"
   },
   "pallet":"staking",
   "palletIndex":"8",
   "storageItem":"nominators",
   "key1":"Ckus4fsdJHxVg69FMDPVbqre3cLUc5R6fWxkMR1HuyQC1Dk",
   "value":{
      "targets":[
         "DSpbbk6HKKyS78c4KDLSxCetqbwnsemv2iocVXwNe2FAvWC",
         "GTzRQPzkcuynHgkEHhsPBFpKdh4sAacVRsnd8vYfPpTMeEY",
         "CoqysGbay3t3Q7hXgEmGJJquhYYpo8PqLwvW1WsUwR7KvXm",
         "DTLcUu92NoQw4gg6VmNgXeYQiNywDhfYMQBPYg2Y1W6AkJF"
      ],
      "submittedIn":"1297",
      "suppressed":false
   }
}
```

File: api/lib/block-listeners/polkadot-node-subscription.js [hunk 2]
```diff
@@ -16,15 +19,65 @@ class PolkadotNodeSubscription extends BaseNodeSubscription {
...
+        data: {
+          sessionIndex: block.sessionIndex
+        }
+      })
+      console.log(
+        `\x1b[36mCurrent session index is ${block.sessionIndex}, fetching validators!\x1b[0m`
+      )
+      const [validators, era] = await Promise.all([
+        dataSource.getAllValidators(),
+        this.store.polkadotRPC.query.staking.activeEra().then(async (era) => {
```
Comment (faboweb) [inline api/lib/block-listeners/polkadot-node-subscription.js]: those should probably become getters in polkadot source

Comment (mariopino) [reply api/lib/block-listeners/polkadot-node-subscription.js]: Done!

File: api/lib/reducers/polkadotV0-reducers.js [hunk 3]
```diff
@@ -218,6 +206,8 @@ function getMessageType(section, method) {
   switch (`${section}.${method}`) {
     case 'balances.transfer':
       return lunieMessageTypes.SEND
+    case 'balances.transferKeepAlive':
```
Comment (faboweb) [inline api/lib/reducers/polkadotV0-reducers.js:209]: can we query for past transactions now?

File: api/lib/source/polkadotV0-source.js [hunk 4]
```diff
@@ -24,22 +29,61 @@ class polkadotAPI {
...
+    // manually set cache to checking it
+    this.cache = new InMemoryLRUCache()
+    this.httpCache = new HTTPCache(this.cache, this.httpFetch)
+  }
+
   setReducers() {
     this.reducers = require('../reducers/polkadotV0-reducers')
   }
 
+  async getRetry(url, intent = 0) {
```
Comment (faboweb) [inline api/lib/source/polkadotV0-source.js:43]: refactor and put into a common lib as used by cosmos and polkadot

File: api/lib/source/polkadotV0-source.js [hunk 5]
```diff
@@ -24,22 +29,61 @@ class polkadotAPI {
...
     const api = this.store.polkadotRPC
     await api.isReady
     return api
   }
 
-  async getNetworkAccountInfo(address, api) {
+  async getNetworkAccountInfo(address) {
     if (typeof address === `object`) address = address.toHuman()
     if (this.store.identities[address]) return this.store.identities[address]
+    // TODO: We are not handling sub-identities
```
Comment (faboweb) [inline api/lib/source/polkadotV0-source.js:82]: maybe fine for now until people complain

**Comment by faboweb** [comment]
Looking awesome! @Bitcoinera can you test this and do a quick comparison on API boot time (until polkadot mainnet is ready) and on proposal overview?

File: api/lib/source/polkadotV0-source.js (hunk 4 above)
Comment (mariopino) [reply api/lib/source/polkadotV0-source.js:43]: I had problems using apollo RESTdatasource functions (this.get, this.cache, this.memoized) in a external function. I tried to move cache flush logic and clear memoized results to `query()` function and pass this.get func as parameter but I got the error `Cannot read property 'fetch'  of undefined`. Any ideas?

**Comment by mariopino** [comment]
Hey @Bitcoinera, I found a couple of issues that I'm fixing rn,  I'll update you when fixed.

File: api/lib/reducers/polkadotV0-reducers.js (hunk 3 above)
Comment (mariopino) [reply api/lib/reducers/polkadotV0-reducers.js:209]: > Hey @Bitcoinera, I found a couple of issues that I'm fixing rn, I'll update you when fixed.

Done @Bitcoinera

Comment (mariopino) [reply api/lib/reducers/polkadotV0-reducers.js:209]: Nop yet :-/

**Comment by mariopino** [comment]
> Hey @Bitcoinera, I found a couple of issues that I'm fixing rn, I'll update you when fixed.

Fixed!

**Comment by Bitcoinera** [comment]
OK, cool. Going to test this now

**Comment by Bitcoinera** [comment]
> @Bitcoinera can you test this and do a quick comparison on API boot time (until polkadot mainnet is ready) and on proposal overview?

API Launch until the "Polkadot is ready" message:

- on develop:

image


- on this branch:

image


Nearly 30 seconds less.

Profilers:

- on develop:

image


- on this branch:

image

**Comment by Bitcoinera** [comment]
I also tested proposals overview and develop tood over 60 seconds to display it, whereas this PR took around 20 seconds.

So I think it is clear we should merge this ASAP :woman_dancing: Great work!

**Review by Bitcoinera — APPROVED** [review]

**Comment by mariopino** [comment]
> I also tested proposals overview and develop tood over 60 seconds to display it, whereas this PR took around 20 seconds.
> 
> So I think it is clear we should merge this ASAP 💃 Great work!

Thank for testing @Bitcoinera!! Great, we still have some old style queries in proposal functions that can be replaced, so we have margin to increase the speed a bit more :-)

File: api/lib/source/polkadotV0-source.js (hunk 5 above)
Comment (mariopino) [reply api/lib/source/polkadotV0-source.js:83]: Good task for another PR!

**Pull request merged**